typing/typemod.ml

```diff
@@ -2446,27 +2518,31 @@ and type_structure ?(toplevel = false) funct_body anchor env sstr =
         Builtin_attributes.parse_standard_implementation_attributes attr;
         Tstr_attribute attr, [], env
   in
-  let rec type_struct env sstr =
+  let toplevel_sig = Option.value toplevel ~default:[] in
+  let rec type_struct env sstr str_acc sig_acc =
     match sstr with
-    | [] -> ([], [], env)
+    | [] ->
+      let sig_acc =
+        Misc.take sig_acc (List.length sig_acc - List.length toplevel_sig)
```

Comment:
Doing maths on list lengths offends my sensibilities. I think you should just make the recursion have another accumulator, one for using with `include functor` and one for using as the final signature. Alternatively have one accumulator and the toplevel items and then combine them at the point where you find an `include functor`.

Reply:
Yes - all the options seemed bad.  But on reflection, I agree having multiple accumulators is least bad.  I've changed it.